File: apps/api/app/schemas/versions.py
"""Version schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class VersionCreate(BaseModel):
    title: str | None = None
    content_json: dict[str, Any] | None = None
    plain_text: str | None = None
    change_summary: str | None = None


class VersionResponse(BaseModel):
    id: str
    document_id: str
    version_number: int
    user_id: str | None = None
    author_name: str
    title: str
    content_json: dict[str, Any] | None = None
    plain_text: str | None = None
    change_summary: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class VersionDiffItem(BaseModel):
    change_type: str  # 'equal' | 'insert' | 'delete'
    text: str


class VersionDiffResponse(BaseModel):
    v1_id: str
    v2_id: str
    v1_version: int
    v2_version: int
    diff_summary: str
    diff_items: list[VersionDiffItem]

File: apps/web/src/lib/api/versions.ts
import { request } from './client';
import type { JSONContent } from '@tiptap/core';

export interface DocumentVersionDTO {
  id: string;
  document_id: string;
  version_number: number;
  user_id?: string | null;
  author_name: string;
  title: string;
  content_json?: JSONContent | Record<string, unknown> | null;
  plain_text?: string | null;
  change_summary?: string | null;
  created_at: string;
}

export interface VersionDiffDTO {
  v1_id: string;
  v2_id: string;
  v1_version: number;
  v2_version: number;
  diff_summary: string;
  diff_items: Array<{ change_type: string; text: string }>;
}

export const versionsApi = {
  list: (documentId: string) => request<DocumentVersionDTO[]>(`/documents/${documentId}/versions`),
  create: (
    documentId: string,
    data: {
      title?: string;
      content_json?: JSONContent | Record<string, unknown>;
      plain_text?: string;
      change_summary?: string;
    }
  ) =>
    request<DocumentVersionDTO>(`/documents/${documentId}/versions`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  get: (documentId: string, versionId: string) =>
    request<DocumentVersionDTO>(`/documents/${documentId}/versions/${versionId}`),
  restore: (documentId: string, versionId: string) =>
    request<DocumentVersionDTO>(`/documents/${documentId}/versions/${versionId}/restore`, {
      method: 'POST',
    }),
  diff: (documentId: string, v1Id: string, v2Id: string) =>
    request<VersionDiffDTO>(`/documents/${documentId}/versions/${v1Id}/diff/${v2Id}`),
};
